module/mempool/stdmap/backdata/heropool/pool.go

```diff
@@ -0,0 +1,300 @@
+
+	e.size++
+	return entityIndex
+}
+
+// Get returns entity corresponding to the entity index from the underlying list.
+func (e Pool) Get(entityIndex EIndex) (flow.Identifier, flow.Entity, uint64) {
+	return e.values[entityIndex].id, e.values[entityIndex].entity, e.values[entityIndex].owner
+}
+
+// sliceIndexForEntity returns a slice index which hosts the next entity to be added to the list.
+func (e *Pool) sliceIndexForEntity() EIndex {
+	if e.free.head.isUndefined() {
+		// the free list is empty, so we are out of space, and we need to eject.
+		if e.ejectionMode == RandomEjection {
+			// turning a random entity into a free head.
+			e.invalidateRandomEntity()
+		} else {
+			// LRU ejection
+			// the used head is the oldest entity, so we turn the used head to a free head here.
+			e.invalidateUsedHead()
+		}
+	}
+
+	// claiming the head of free list as the slice index for the next entity to be added
+	return e.claimFreeHead()
+}
+
+// Size returns total number of entities that this list maintains.
+func (e Pool) Size() uint32 {
+	return e.size
+}
+
+// getHeads returns entities corresponding to the used and free heads.
+func (e Pool) getHeads() (*cachedEntity, *cachedEntity) {
+	var usedHead, freeHead *cachedEntity
+	if !e.used.head.isUndefined() {
+		usedHead = &e.values[e.used.head.sliceIndex()]
+	}
+
+	if !e.free.head.isUndefined() {
+		freeHead = &e.values[e.free.head.sliceIndex()]
+	}
+
+	return usedHead, freeHead
+}
+
+// getTails returns entities corresponding to the used and free tails.
+func (e Pool) getTails() (*cachedEntity, *cachedEntity) {
+	var usedTail, freeTail *cachedEntity
+	if !e.used.tail.isUndefined() {
+		usedTail = &e.values[e.used.tail.sliceIndex()]
+	}
+
+	if !e.free.tail.isUndefined() {
+		freeTail = &e.values[e.free.tail.sliceIndex()]
+	}
+
+	return usedTail, freeTail
+}
+
+// connect links the prev and next nodes as the adjacent nodes in the double-linked list.
+func (e *Pool) connect(prev poolIndex, next EIndex) {
+	e.values[prev.sliceIndex()].node.next.setPoolIndex(next)
+	e.values[next].node.prev = prev
+}
+
+// invalidateUsedHead moves current used head forward by one node. It
+// also removes the entity the invalidated head is presenting and appends the
+// node represented by the used head to the tail of the free list.
+func (e *Pool) invalidateUsedHead() EIndex {
+	headSliceIndex := e.used.head.sliceIndex()
+	e.invalidateEntityAtIndex(headSliceIndex)
+
+	return headSliceIndex
+}
+
+// invalidateRandomEntity invalidates a random node from the used linked list, and appends it to the tail of the free list.
+// It also removes the entity that the invalidated node is presenting.
+func (e *Pool) invalidateRandomEntity() EIndex {
+	// in order not to keep failing on finding a random valid node to invalidate,
+	// we only try a limited number of times, and if we fail all, we invalidate the used head.
+	var index = e.used.head.sliceIndex()
+
+	for i := 0; i < maximumRandomTrials; i++ {
```

Comment:
I'm a bit confused about the `maximumRandomTrials` and this algorithm:

Surely we only need to call this function IF the free list contains zero entries? (otherwise we would just take the next entry from the free list, or?) Which also means that the pool is completely full, which also means that we should ALWAYS hit a random pool index which is valid (because they are all used...), or?

Is this algorithm being used because we are not keeping count of the number of entries in the free list? In which case, why not keep count and eliminate the unnecessary `maximumRandomTrials`?

Reply:
> Surely we only need to call this function IF the free list contains zero entries? (otherwise we would just take the next entry from the free list, or?) Which also means that the pool is completely full, which also means that we should ALWAYS hit a random pool index which is valid (because they are all used...), or?

True

> Is this algorithm being used because we are not keeping count of the number of entries in the free list? In which case, why not keep count and eliminate the unnecessary maximumRandomTrials?

No, the purpose of `maximumRandomTrials` is two folds: 
- To make the method of `invalidateRandomEntity` safe for the evolved use cases. Currently, we only use this method when we are at capacity, and we need to eject one element at random. However, if this method is ever will be used to perform _batch random ejection_ (as we have in the normal map-based backend), the performance is exceedingly degraded as the size of the batch gets bigger. The reason is the distribution of used entities gets scattered with bigger batch sizes. 

- As a side benefit, having `maximumRandomTrials` helps extending the testing coverage of `HeroPool` with more advance cases like [testing random invalidation](https://github.com/onflow/flow-go/blob/c5baa0ea1d8fa403fde197c60fb17460572133b6/module/mempool/stdmap/backdata/herocache/heropool/pool_test.go#L125-L193).

Reply:
@yhassanzadeh13 thanks for the detailed comment! :-)

> if this method is ever will be used to perform batch random ejection (as we have in the normal map-based backend), the performance is exceedingly degraded as the size of the batch gets bigger.

AFAIK @jwinkler2083233 introduced the batch random ejection purely as a (better than nothing but still far from optimal) performance enhancement because the original single random ejection walked (which is very slow) a random number of keys to delete one, i.e. the batch random ejection was only introduced because nobody could think of a better way to increase ejection performance at the time.

Is it worth it to introduce this low performing heroCache `maximumRandomTrials` mechanism just for backwards compatibility with the batch random ejection mechanism which was never really wanted in the first place? :-)

AFAIK the batch random ejection mechanism is not required by any other biz logic anywhere. So here are some suggestions for a workaround:

1. Rig the code so that trying to use batch random ejection with heroCache will panic, i.e. in case somebody tries to use / configure it in the future then it won't work.

2. Batch random ejection can still work if:

(a) the batch random ejection is only performed when heroCache is full, and

(b) if the random numbers are all calculated in advance of deletion and there are no collisions, i.e. the random numbers are all unique.

AFAIK this is also how the batch random ejection algorithm introduced by @jwinkler2083233 also works. It walks a random number of keys, deletes, then walks another random number of keys, deletes, and so on, i.e. the random keys never collide. This is a lot different from the algorithm here under review which chooses a random index, tries to delete, choose another random index (which maybe collides with a previously deleted index), tries to delete, etc.

In theory the walk and delete strategy could work exactly the same way for heroCache but instead of walking keys in a hash table, the walking can be achieved by adding a random value to an index and walking the heroCache slice?

Whether option 1 or 2 then `maximumRandomTrials` can be refactored away, or?

Reply:
fixed in https://github.com/onflow/flow-go/pull/1778/commits/74aef8dce1c2683d1caf122a3635bf244ebfdff8